### PR TITLE
imp: RPC server refactor

### PR DESCRIPTION
**Pull request by facs95**
# Description
The PR is just the beginning of the major refactor of the whole backend but hopefully it serves as a template for us to progressively migrate the legacy code. Main improvements are:

- Logging
- Proper encoding and decoding of json objects
- Code maintainability
- Consistent responses
- Restfull structure of our endpoints

I will try to explain the changes I did and the reasoning for it, feel free to split this PR into a set of smaller ones if necessary.

- `cmd/server`
    
    Decided to move our `main.go` file into a `/cmd` directory because we are managing different things from the same repo (cron, redis, etc..). This way is more clear what we are building. To build the server binary run:
    
      
    
    ```go
    go build ./cmd/server/ 
    ```
    
    I updated `Dockerfile` with the necessary change to build.
    
- `internal/server`
    
    It manages everything related to the RPC server

- `internal/config`
    
    Handles configuration file of the server. At the moment is very simple but the idea is to extend this to a proper server configuration file. I made it an independent package from the `rpc` for it to be global config file (logging,  ws, etc..). 

- `internal/handler`
    
    Handles the endpoints that we support. 
    
    It has a `v1` and `v2` package, each with its own handler. 
    
    - `v1` - Has all the legacy endpoints that we supported with the previous structure. This endpoints will be progressively depecrated. **Files in this packaged were not modified, only moved from their previous package. There is no need to spend a lot of time reviewing them.**
    - `v2`  - New structure of the endpoints. With some important things I wanted to note
        - Routes have a RESTfull structure path.
        - We currently manage logging thorugh the `ctx.logger` provided by `fasthttp`. We could probably later change this for a global logger passed by the server that can be modified through config (loglevel, logfile, etc…). But we should have
        - Every endpoint / handler should be a method of the v2 handler.
        - The Handler struct gives access to the `NumiaClient` for every handler to be able to access it and make the necessary queries. This helps maintaining code clean and helps with separation of concerns.
        - Responses have consistent structure and status codes. This can be found on the `responses.go` file.
        - Error responses always return a message that is displayable in the UI. If an unknown internal error occurred we return a default message `Something went wrong..` and log the necessary information. The error response always has the structure:
            
            ```json
            {
            	"error": "Some error message..."
            }
            ```
            
- `internal/numia`
    
    The package provides a `Client` to make HTTP requests to the Numia API. It is used by the handler.

## New Endpoints
The PR also includes the addition of 3 new endpoints as the beginning of the Numia migration:
- `/v2/rewards/${address}`
- `/v2/delegations/${address}`
- `/v2/height`
    

# Important

It requires some infra changes as we now need the following Environment Variables:

- `NUMIA_API_KEY` - Required
- `NUMIA_RPC_ENDPOINT` - Required
- `RPC_SERVER_PORT` - optional

# Todos

- We need tests for at least basic functionality
- Add support for swagger
- Extend server configuration as necessary
- Migrate endpoints to v2 structure to fully deprecate v1
- Add gosec

### CODE REVIEW

**Review by fedekunze — APPROVED** [review]
Great improvements! Let's add a changelog and the license comments and the we can address the TODOS and further improvements in a separate PR

File: internal/handler/v2/delegations.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	if address == "" {
+		ctx.SetStatusCode(fasthttp.StatusBadRequest)
+		errorResponse := &ErrorResponse{
+			Error: "Missing address",
+		}
+		sendJSONResponse(ctx, errorResponse)
+		return
+	}
+
+	delegations, err := h.numiaRPCClient.QueryDelegations(address)
```
Comment (GAtom22) [inline internal/handler/v2/delegations.go:38]: Q: should we retry the query if fails? maybe we can use a retryable HTTP client, e.g. [this lib](https://github.com/hashicorp/go-retryablehttp), or add retry on our existing code

File: internal/numia/client.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		return fmt.Errorf("error creating request: %s", err.Error())
+	}
+
+	// Set authorization header
+	authHeader := fmt.Sprintf("Bearer %s", c.apiKey)
+	req.Header.Set("Authorization", authHeader)
+	req.Header.Set("Accept", "application/json")
+
+	// Send HTTP request
+	client := &http.Client{}
```
Comment (GAtom22) [inline internal/numia/client.go:48]: ditto here, should we add retries ([possible lib](https://github.com/hashicorp/go-retryablehttp))?

**Review by GAtom22 — APPROVED** [review]
LGTM! Great work @facs95 !! Left only one question regarding requests retries to numia

File: internal/handler/v2/delegations.go (hunk 1 above)
Comment (facs95) [reply internal/handler/v2/delegations.go:38]: I think thats a great idea! Can you create a ticket and assign to it to me? I can tackle that on a separate PR once back

Comment (GAtom22) [reply internal/handler/v2/delegations.go:38]: created a ticket for this

**Pull request merged**